### ISSUE TITLE
False-positive Lint/OrderedMagicComments warning

### ISSUE THREAD

**Issue by DmitryPogrebnoy** [rubocop/rubocop#12213]
There is a specific case when the Lint/OrderedMagicComments cop produces a false-positive warning.

Here is an example:

```
# frozen_string_literal: true

# Example:
#
#   re = eval("# encoding: ISO-8859-1\n/foo \\xff?/")
#   re.encoding                 # => #<Encoding:ISO-8859-1>
#   re =~ "foo".encode("UTF-8") # => 0
#   re =~ "foo\u0100"           # Raises Encoding::CompatibilityError

```

## Expected behavior

There is no warning.

## Actual behavior

The Lint/OrderedMagicComments warning is present.

Here is an output `rubocop --debug`:
```
For /Users/dmitry.pogrebnoy/RubymineProjects/untitled25: Default configuration from /Users/dmitry.pogrebnoy/.rvm/gems/ruby-3.3.0-preview2-debug/gems/rubocop-1.56.3/config/default.yml
Use parallel by default.
Skipping parallel inspection: only a single file needs inspection
Inspecting 1 file
Scanning /Users/dmitry.pogrebnoy/RubymineProjects/untitled25/1.rb
W

Offenses:

1.rb:5:1: W: [Correctable] Lint/OrderedMagicComments: The encoding magic comment should precede all other magic comments.
#   re = eval("# encoding: ISO-8859-1\n/foo \\xff?/")
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1 file inspected, 1 offense detected, 1 offense autocorrectable
Finished in 0.18397099990397692 seconds
```

## Steps to reproduce the problem

1. Create empty ruby file
2. Copy following example to file.
```
# frozen_string_literal: true

# Example:
#
#   re = eval("# encoding: ISO-8859-1\n/foo \\xff?/")
#   re.encoding                 # => #<Encoding:ISO-8859-1>
#   re =~ "foo".encode("UTF-8") # => 0
#   re =~ "foo\u0100"           # Raises Encoding::CompatibilityError
```
3. Run rubocop on that file.

## RuboCop version

1.56.3 (using Parser 3.2.2.3, rubocop-ast 1.29.0, running on ruby 3.3.0) [arm64-darwin22]
